quic/transport/ngtcp2/connection/drainingstate.nim

```diff
@@ -0,0 +1,67 @@
+import pkg/chronos
+import ../../../udp/datagram
+import ../../quicconnection
+import ../../connectionid
+import ../../stream
+import ../../timeout
+import ./disconnectingstate
+import ./closedstate
+
+type
+  DrainingConnection* = ref object of ConnectionState
+    connection*: QuicConnection
+    ids: seq[ConnectionId]
+    timeout: Timeout
+    duration: Duration
+    done: AsyncEvent
+
+proc init*(state: DrainingConnection,
+           ids: seq[ConnectionId], duration: Duration) =
+  state.ids = ids
+  state.duration = duration
+  state.done = newAsyncEvent()
+
+proc newDrainingConnection*(ids: seq[ConnectionId],
+                            duration: Duration): DrainingConnection =
+  let state = DrainingConnection()
+  state.init(ids, duration)
+  state
+
+proc onTimeout(state: DrainingConnection) =
+  state.done.fire()
+
+{.push locks: "unknown".}
+
+method enter*(state: DrainingConnection, connection: QuicConnection) =
+  state.connection = connection
```

Comment:
I think there should be an assert state.connection = nil

Reply:
I added protection against state reentry in 923ad8d.